### PR TITLE
Added L0 test cases for kubernetes

### PR DESCRIPTION
**Pull request by Anumita**

### CODE REVIEW

File: Tasks/Kubernetes/Tests/TestSetup.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+tr.setAnswers(<any>a);
+
+// Create mock for fs module
+let fs = require('fs');
+let fsClone = Object.assign({}, fs);
+fsClone.existsSync = function(filePath) {
+    switch (filePath) {
+        case "kubectl":
+            if(JSON.parse(process.env[shared.isKubectlPresentOnMachine]))
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/TestSetup.ts]: why JSON.parse

File: Tasks/Kubernetes/Tests/L0.ts [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.stdout.indexOf(`[command]${shared.formatPath("newUserDir/kubectl.exe")} --kubeconfig ${shared.formatPath("newUserDir/config")} get pods`) != -1, "kubectl get should run");
+        console.log(tr.stderr);
+        done();
+    });
+
+    it('Run sucessfully when the user provides the version for kubectl with checkLatest as false', (done:MochaDone) => {
+        let tp = path.join(__dirname, 'TestSetup.js');
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: There is no check, in the test that kubectl is downloaded and version 1.7 is downloaded.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        process.env[shared.TestEnvVars.versionOrLocation] = "version";
+        process.env[shared.TestEnvVars.versionSpec] = "1.7.0";
+        process.env[shared.TestEnvVars.checkLatest] = "true";
+        process.env[shared.isKubectlPresentOnMachine] = "false";
+        tr.run();
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]${shared.formatPath("newUserDir/kubectl.exe")} --kubeconfig ${shared.formatPath("newUserDir/config")} get pods`) != -1, "kubectl apply should run");
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: command you are trying to run is get pods not apply. Please fix the message.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        console.log(tr.stderr);
+        done();
+    });
+
+    it('Run sucessfully when the user provides the version for kubectl with checkLatest as true', (done:MochaDone) => {
+        let tp = path.join(__dirname, 'TestSetup.js');
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: I can not see any check that latest version is looked for and downloded.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        process.env[shared.TestEnvVars.arguments] = "pods";    
+        process.env[shared.TestEnvVars.versionOrLocation] = "version";
+        process.env[shared.TestEnvVars.versionSpec] = "1.7.0";
+        process.env[shared.isKubectlPresentOnMachine] = "false";
+        tr.run();
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]${shared.formatPath("newUserDir/kubectl.exe")} --kubeconfig ${shared.formatPath("newUserDir/config")} get pods`) != -1, "kubectl apply should run");
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: command is get pods and message is apply. This error is in all test please fix it.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        console.log(tr.stderr);
+        done();
+    });
+
+    it('Run sucessfully when the user provides the version for kubectl with checkLatest as false and version that dosent have a v prefix ', (done:MochaDone) => {
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: Duplicate test. Run successfully when the user provides the version for kubectl with check latest as false

File: Tasks/Kubernetes/Tests/L0.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        delete process.env[shared.TestEnvVars.useConfigurationFile];
+        delete process.env[shared.TestEnvVars.secretName];
+        delete process.env[shared.TestEnvVars.forceUpdate];
+        delete process.env[shared.TestEnvVars.outputFormat];
+        delete process.env[shared.TestEnvVars.kubectlOutput];
+    });
+    after(function () {
+    });
+
+    it('Run sucessfully when the user provides a specific location for kubectl', (done:MochaDone) => {
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: type in sucessfully => successfully in every test.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        process.env[shared.TestEnvVars.arguments] = "pods";    
+        process.env[shared.TestEnvVars.versionOrLocation] = "version";
+        process.env[shared.TestEnvVars.versionSpec] = "1.7";
+        process.env[shared.isKubectlPresentOnMachine] = "false";
+        tr.run();
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]${shared.formatPath("newUserDir/kubectl.exe")} --kubeconfig ${shared.formatPath("newUserDir/config")} get pods`) != -1, "kubectl apply should run");
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: I think you copied the checks in output section as is in all test cases. There is no check to validate that, tests are running the way they are expected. Tests might succeed but does not execute the way they are supposed to.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]kubectl --kubeconfig ${shared.formatPath("newUserDir/config")} delete pod -all`) != -1, "kubectl delete should run");
+        console.log(tr.stderr);
+        done();
+    });
+
+    it('Runs successfully for kubectl exec', (done:MochaDone) => {
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: We dont have to add test cases for each command.

File: Tasks/Kubernetes/Tests/L0.ts [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]kubectl --kubeconfig ${shared.formatPath("newUserDir/config")} set env pods --all --list`) != -1, "kubectl set should run");
+        console.log(tr.stderr);
+        done();
+    });
+
+    it('Runs successfully for kubectl top', (done:MochaDone) => {
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/L0.ts]: same comment, tests for each command is duplicate.

File: Tasks/Kubernetes/Tests/TestSetup.ts [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        return new Date(1996,3,5);
+    },
+    getNewUserDirPath: function() {
+        return newUserDirPath;
+    },
+    getStableKubectlVersion: function(){
+        return "v1.6.6";
+    },
+    getKubectlVersion: async function(versionSpec, checkLatest) : Promise<string> {
+        let version: string = "v1.6.6";
```
Comment (jikuma) [inline Tasks/Kubernetes/Tests/TestSetup.ts]: We should avoid having logic in mocq.

File: Tasks/Kubernetes/Tests/L0.ts (hunk 7 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: done

File: Tasks/Kubernetes/Tests/L0.ts (hunk 5 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: done

File: Tasks/Kubernetes/Tests/L0.ts (hunk 2 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: applied check

File: Tasks/Kubernetes/Tests/L0.ts (hunk 3 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: Fixed message

File: Tasks/Kubernetes/Tests/L0.ts (hunk 4 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: Added check

File: Tasks/Kubernetes/Tests/L0.ts (hunk 9 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: Removed duplicate test cases

File: Tasks/Kubernetes/Tests/L0.ts (hunk 10 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: Removed duplicate test cases

File: Tasks/Kubernetes/Tests/TestSetup.ts (hunk 11 above)
Comment (Anumita) [reply Tasks/Kubernetes/Tests/TestSetup.ts]: Need to add the logic as the mock getStableKubectlVersion() will not be called if we use the original getKubectlVersion().

File: Tasks/Kubernetes/Tests/L0.ts [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        process.env[shared.TestEnvVars.command] = shared.Commands.get;
+        process.env[shared.TestEnvVars.arguments] = "pods";
+        process.env[shared.TestEnvVars.containerType] = shared.ContainerTypes.ContainerRegistry;
+        process.env[shared.TestEnvVars.secretName] = "my-secret";
+        tr.run();
+
+        assert(tr.invokedToolCount == 2, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]kubectl --kubeconfig ${shared.formatPath("newUserDir/config")} create secret docker-registry my-secret --docker-server=https://index.docker.io/v1/ --docker-username=test --docker-password=regpassword --docker-email=test@microsoft.com`) != -1, "kubectl create should run");
```
[the remark this answers is not in the document]
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: added check for delete secret

File: Tasks/Kubernetes/Tests/L0.ts [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        process.env[shared.TestEnvVars.command] = shared.Commands.get;
+        process.env[shared.TestEnvVars.arguments] = "pods";
+        process.env[shared.TestEnvVars.containerType] = shared.ContainerTypes.AzureContainerRegistry;
+        process.env[shared.TestEnvVars.secretName] = "my-secret";
+        tr.run();
+
+        assert(tr.invokedToolCount == 2, 'should have invoked tool one times. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]kubectl --kubeconfig ${shared.formatPath("newUserDir/config")} create secret docker-registry my-secret --docker-server=ajgtestacr1.azurecr.io --docker-username=spId --docker-password=spKey --docker-email=ServicePrincipal@AzureRM`) != -1, "kubectl create should run");
```
[the remark this answers is not in the document]
Comment (Anumita) [reply Tasks/Kubernetes/Tests/L0.ts]: added

**Pull request merged**